jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -115,14 +142,49 @@ public Builder addDirectoryContents(
         Predicate<Path> pathFilter,
         AbsoluteUnixPath basePathInContainer)
         throws IOException {
+      return addDirectoryContents(
+          layerType, sourceRoot, pathFilter, basePathInContainer, ImmutableMap.of());
+    }
+
+    /**
+     * Adds directory contents to a layer selectively (via {@code pathFilter}) and recursively.
+     * {@code sourceRoot} must be a directory. Empty directories will always be added regardless of
+     * {@code pathFilter}, except that {@code sourceRoot} is never added. Permissions are specified
+     * via {@code permissionsMap}, which maps from extraction path on the container to file
+     * permissions.
+     *
+     * @param layerType the layer to add files into
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter filter that determines which files (not directories) should be added
+     * @param basePathInContainer directory in the layer into which the source contents are added
+     * @param permissionsMap the map from absolute path on container to file permission
+     * @return this
+     * @throws IOException error while listing directories
+     * @throws NotDirectoryException if {@code sourceRoot} is not a directory
+     */
+    // TODO: Use in plugins
+    public Builder addDirectoryContents(
+        LayerType layerType,
+        Path sourceRoot,
+        Predicate<Path> pathFilter,
+        AbsoluteUnixPath basePathInContainer,
+        Map<AbsoluteUnixPath, Set<PosixFilePermission>> permissionsMap)
```

Comment:
Wondering if it might not be more flexible to encode this instead as a `Function<AbsoluteUnixPath,Set<PosixFilePermission>>`?  (Or a `BiFunction` with the actual on-disk path?)

Reply:
Not sure what you mean, can you be more specific?

Reply:
I believe what @briandealwis said is like a map is basically a function. That is, `map.put(path, permissions)` is like `Function<...> f = key -> key == path ? permissions : ...`. Conceptually they are somewhat interchangeable, but I'm not opinionated in which one is better here. But if you have a map, wrapping it as a function is probably very easy, which is just `Function f = map::get`.

Reply:
It adds flexibility for the future.  I'm thinking that if we may already have a Maven or Gradle model object representing the mapping of files to permissions.  Forcing the building of a map may be unnecessary.  And it simplifies adding rules like _all .class files should always be 0444_.

Reply:
I'll change this when I get to the plugins if need be.